### PR TITLE
Add a verify method which also takes in the uncompressed public key.

### PR DESCRIPTION
**Pull request by ValarDragon**
This is useful for the proof of stake use case, since we often validate multiple signatures from the same public key. Based off of the numbers in the official Ed25519 paper, the function with the uncompressed pubkey should save about 10% of the computation time.

### CODE REVIEW

File: ed25519.go [hunk 1]
```diff
@@ -124,11 +134,18 @@ func Verify(publicKey *[PublicKeySize]byte, message []byte, sig *[SignatureSize]
...
 	var R edwards25519.ProjectiveGroupElement
-	var b [32]byte
-	copy(b[:], sig[32:])
-	edwards25519.GeDoubleScalarMultVartime(&R, &hReduced, &A, &b)
+	var s [32]byte
+	copy(s[:], sig[32:])
+
+	// https://tools.ietf.org/html/rfc8032#section-5.1.7 requires that s be in
+	// the range [0, order) in order to prevent signature malleability.
+	if !edwards25519.ScMinimal(&s) {
```
Comment (liamsi) [inline ed25519.go:142]: This also contains the diff for the upstream changes from x/crypto? I think we should merge this in a separate PR first and add the `VerifyUncompressedKey` here only (by rebasing / merging).

Comment (ValarDragon) [reply ed25519.go:142]: Okay!

Comment (liamsi) [reply ed25519.go:142]: You can keep the PR open. I'll merge #6 after making sure the diff exactly matches the changes on x/crypto. Then, you can rebase on the branch that contains the merge.

File: ed25519.go [hunk 2]
```diff
@@ -110,6 +110,16 @@ func Verify(publicKey *[PublicKeySize]byte, message []byte, sig *[SignatureSize]
 	if !A.FromBytes(publicKey) {
 		return false
 	}
+	return VerifyUncompressedKey(publicKey, &A, message, sig)
+}
+
+// VerifyUncompressedKey returns true iff sig is a valid signature of message by publicKey.
+// This takes in the uncompressed form of the public key (A) to avoid computing that internally.
+func VerifyUncompressedKey(publicKey *[PublicKeySize]byte, A *edwards25519.ExtendedGroupElement, message []byte, sig *[SignatureSize]byte) bool {
```
Comment (liamsi) [inline ed25519.go:118]: If we introduce a new public method, we should have a test for that.

Comment (ValarDragon) [reply ed25519.go:118]: How would such a test differ from the normal verify though? I would basically be replicating the code inside of Verify, and I'm not sure how useful that is as a test. (Uncompress the pubkeys using the same decompression method, and then run the same test vectors, which is exactly whats done by testing Verify)

Comment (liamsi) [reply ed25519.go:118]: Tests also serve as a form of documentation here. One can look into the test and see the exact behaviour you just described above. Does that make sense to you?

**Comment by liamsi** [comment]
> Based off of the numbers in the official Ed25519 paper, the function with the uncompressed pubkey should save about 10% of the computation time.

That is pretty cool! 👍 We need to make sure that this 10% improvement, actually leads to a measurable improvement in a real deployment and hence is really worth exposing some edwards25519 compression [interna](https://github.com/tendermint/ed25519/blob/d8387025d2b9d158cf4efb07e7ebf814bcce2057/edwards25519/edwards25519.go#L719-L763) (and later needing to change the [keybase interface](https://github.com/tendermint/go-crypto/blob/1a2f468695f9a57907cda7f018855ea7cc912175/keys/types.go#L8-L35) to provide the actual cached uncompressed public key.

Would you like to do some perf measurements to compare how much of an improvement this would actually bring in tendermint/cosmos deployment? I would suggest not to merge before that.

**Pull request closed without merging**